pkg/operator/ceph/object/notification/obc_label_controller.go

```diff
@@ -0,0 +1,216 @@
+	"github.com/rook/rook/pkg/operator/ceph/object/topic"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	notificationLabelPrefix = "bucket-notification-"
+)
+
+// ReconcileOBCLabels reconciles a ObjectBucketClaim labels
+type ReconcileOBCLabels struct {
+	client           client.Client
+	context          *clusterd.Context
+	clusterInfo      *cephclient.ClusterInfo
+	clusterSpec      *cephv1.ClusterSpec
+	opManagerContext context.Context
+}
+
+func addOBCLabelReconciler(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+	logger.Info("successfully started")
+
+	// Watch for changes on the OBC CRD object
+	err = c.Watch(&source.Kind{Type: &bktv1alpha1.ObjectBucketClaim{}}, &handler.EnqueueRequestForObject{}, opcontroller.WatchControllerPredicate())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the cluster for a ObjectBucketClaim object and makes changes based on the state read
+// and the ObjectBucketClaim labels
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileOBCLabels) Reconcile(context context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// workaround because the rook logging mechanism is not compatible with the controller-runtime logging interface
+	reconcileResponse, err := r.reconcile(request)
+	if err != nil {
+		logger.Errorf("failed to reconcile %v", err)
+	}
+
+	return reconcileResponse, err
+}
+
+func (r *ReconcileOBCLabels) reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the ObjectBucketClaim instance
+	obc := bktv1alpha1.ObjectBucketClaim{}
+	err := r.client.Get(r.opManagerContext, request.NamespacedName, &obc)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Debugf("ObjectBucketClaim %q resource not found. Ignoring since resource must be deleted.", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errors.Wrapf(err, "failed to retrieve ObjectBucketClaim %q", request.NamespacedName)
+	}
+
+	// if ObjectBucket was created yet, reschedule in 5 seconds
+	if obc.Spec.ObjectBucketName == "" {
+		logger.Infof("ObjectBucketClaim %q resource did not create the bucket yet. will retry", request.NamespacedName)
+		return waitForRequeueIfObjectBucketNotReady, nil
+	}
+
```

Comment:
From this point in the obc label reconciler, I would suggest that this method should look at the labels on the OBC and determine which CephBucketNotifications to look for. List those notifications, filtering out ones that don't have notification labels, and then call the reusable function in a loop to provision each notification. Doing this might be a little bit less efficient, but it will be much easier to test and maintain.